src/combinatorics/hindman.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2021 David Wärn. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Wärn
+-/
+import topology.stone_cech
+import topology.algebra.semigroup
+import data.stream
+
+/-!
+# Hindman's theorem on finite sums
+
+We prove Hindman's theorem on finite sums, using idempotent ultrafilters.
+
+Given an infinite sequence `a₀, a₁, a₂, …` of positive integers, the set `FS(a₀, …)` is the set
+of positive integers that can be expressed as a finite sum of `aᵢ`'s, without repetition. Hindman's
+theorem asserts that whenever the positive integers are finitely colored, there exists a sequence
+`a₀, a₁, a₂, …` such that `FS(a₀, …)` is monochromatic. There is also a stronger version, saying
+that whenever a set of the form `FS(a₀, …)` is finitely colored, there exists a sequence
+`b₀, b₁, b₂, …` such that `FS(b₀, …)` is monochromatic and contained in `FS(a₀, …)`. We prove both
+these versions for a general semigroup `M` instead of `ℕ+` since it is no harder, although this
+special case implies the general case.
+
+The idea of the proof is to extend the addition `(+) : M → M → M` to addition `(+) : βM → βM → βM`
+on the space `βM` of ultrafilters on `M`. One can prove that if `U` is an _idempotent_ ultrafilter,
+i.e. `U + U = U`, then any `U`-large subset of `M` contains some set `FS(a₀, …)` (see
+`exists_FS_of_large`). And with the help of a general topological argument one can show that any set
+of the form `FS(a₀, …)` is `U`-large according to some idempotent ultrafilter `U` (see
+`exists_idempotent_ultrafilter_le_FS`). This is enough to prove the theorem since in any finite
+partition of a `U`-large set, one of the parts is `U`-large.
+
+## Main results
+
+- `FS_partition_regular`: the strong form of Hindman's theorem
+- `exists_FS_of_finite_cover`: the weak form of Hindman's theorem
+
+## Tags
+
+Ramsey theory, ultrafilter
+
+-/
+
+open filter
+
+@[to_additive]
+instance {M} [has_mul M] : has_mul (ultrafilter M) := { mul := λ U V, (*) <$> U <*> V }
```

Comment:
If you say that these definitions are not well-behaved, please make them definitions, and add the `instance` attribute locally.